### PR TITLE
Use Octopus API instead of calling the octo-cli

### PR DESCRIPTION
**Pull request by RmMsr**
Implementation of https://github.com/kolonialno/velo/issues/154

To test it, please follow the [development guide](https://github.com/kolonialno/velo-action/blob/RmMsr/octopus-api/docs/development.md)

**Disclaimer**: This PR has grown much bigger than intended while learning to develop a testable github action. Besides the core change of removing the octo-cli and addition of a Octopus client it contains:
* Option to specify a duration when waiting for a successful deployment `wait_for_success_seconds`
* Improve validation output for settings
* Improved wording on octopus specific setting. This should not break existing setups. The removed settings are not in use
* Add env setup for local execution
* Fix dockerized execution and settings pass through
* Add documentation how to run the action locally
* Creating a testing decorator to define and assert expected API requests
* Add tests for Settings
* Bump python to 3.10
* Update of dependencies
* Fix of numerous Linter complaints and quality checks

With the amount of changes I do not expect a full code review. But I would like some feedback on the Interface choices of the octopus module classes or the approach to testing.

The Deploy workflow of the [example-deploy-project](https://github.com/kolonialno/example-deploy-project/blob/main/.github/workflows/deploy.yml) is altered to use the velo-action from this branch. The project contains no tenants - otherwise it should be a good test case.

### CODE REVIEW

File: velo_action/action.py [hunk 1]
```diff
@@ -63,7 +63,7 @@ def action(input_args: Settings):
         if not input_args.project:
             raise ValueError("project not specified")
         if not input_args.service_account_key:
-            raise ValueError("gcp service account key not specified")
+            logger.warning("gcp service account key not specified")
 
         g = gcp.GCP(input_args.service_account_key)
         octopus_cli_server = g.lookup_data(
```
Comment (larseen) [inline velo_action/action.py:69]: This is quite wierd? 
Why are we looking up stuff as a secret using GCP?
Is this becasue it is a action and we dont want do require skjer a?crets to be avialable? 
Seems strange that we can not give it settings/configuration from environment

Comment (RmMsr) [reply velo_action/action.py:69]: Good question. I am not sure why exactly. I see two benefits. It's working the same with minimal configuration from any platform (local, kubernetes, Github). And the configuration is centrally managed and versioned. I don't feel much motivation to change it.

Comment (andersliland) [reply velo_action/action.py:69]: The idea is to reduce the inputs to velo-action as much as possible to make it easy to use. 

By only suppling the Google Service Account Key, the action can fetch the rest from Secret Manager.

The alternative is that every velo-action instance in Github Action Workflow must add the Octopus server and API key as secrets from Github...

File: velo_action/action.py [hunk 2]
```diff
@@ -63,7 +63,7 @@ def action(input_args: Settings):
         if not input_args.project:
             raise ValueError("project not specified")
         if not input_args.service_account_key:
-            raise ValueError("gcp service account key not specified")
```
Comment (andersliland) [inline velo_action/action.py:66]: Why change this? It the Service Account is not specified the action should stop and return an exception

File: velo_action/gcp.py [hunk 3]
```diff
@@ -12,25 +13,12 @@
...
+            self._auth_service_account(service_account_key)
         else:
-            service_account_info = json.loads(service_account_key)
-        credentials = service_account.Credentials.from_service_account_info(
-            service_account_info
-        )
-        self.scoped_credentials = credentials.with_scopes(
-            ["https://www.googleapis.com/auth/cloud-platform"]
-        )
+            logger.info("Using local credentials")
```
Comment (andersliland) [inline velo_action/gcp.py:21]: What if no service account key is specified and local credentials dont exist?

We should return a nice error for the Github workflow log...

File: velo_action/octopus_api.py [hunk 4]
```diff
@@ -0,0 +1,264 @@
...
+
+    def _post_request(self, path, data):
+        url = urllib.parse.urljoin(self._baseurl, path)
+        response = requests.post(url, json=data, headers=self._headers)
+        logger.debug(
+            f"{response.request.method} {response.url}: {response.status_code}"
+        )
+        return _handle_response(response)
+
+    def _determine_latest_deploy_packages(self, project_id):
```
Comment (andersliland) [inline velo_action/octopus_api.py:197]: does this allow us to select the velo version to use when creating a release?

File: velo_action/octopus_api.py [hunk 5]
```diff
@@ -0,0 +1,264 @@
...
+
+import requests
+
+logger = logging.getLogger(name="octopus")
+
+_RELEASE_REGEX = r"^(|\+.*)$"
+_MAX_WAIT_TIME = timedelta(minutes=3)
+
+
+class Octopus:
```
Comment (andersliland) [inline velo_action/octopus_api.py:18]: would love some tests on this class!

**Review by andersliland — CHANGES_REQUESTED** [review]
Would love some tests on the Octopus class.

And an update to the docs on how to run the action locally when developing.

How can this easily be tested agains example deploy project for instance?

File: velo_action/octopus_api.py (hunk 5 above)
Comment (RmMsr) [reply velo_action/octopus_api.py:16]: Indeed. We have to figure out what to test. I think the data model between server and client do not provide much value to test. I ll look through the functions and see if there is some logic that I can easily test

File: velo_action/action.py (hunk 1 above)
Comment (RmMsr) [reply velo_action/action.py:71]: By the way. the velo-action can now be ran with your local gcloud credentials. You just need to grant yourself the privileges to access the secrets.

File: velo_action/action.py (hunk 2 above)
Comment (RmMsr) [reply velo_action/action.py:66]: Yes, as the primary use case is still within GitHub workflows. I thought that qualifies for a warning.

File: velo_action/gcp.py (hunk 3 above)
Comment (RmMsr) [reply velo_action/gcp.py:21]: Yes, we can. The 403 you get from the gcp SDK is not so bad though.

Comment (RmMsr) [reply velo_action/gcp.py:21]: That is:

```google.auth.exceptions.DefaultCredentialsError: Could not automatically determine credentials. Please set GOOGLE_APPLICATION_CREDENTIALS or explicitly create credentials and re-run the application. For more information, please see https://cloud.google.com/docs/authentication/getting-started
```

But we should add a note for `INPUT_SERVICE_ACCOUNT_KEY`

File: velo_action/octopus_api.py (hunk 5 above)
Comment (larseen) [reply velo_action/octopus_api.py:16]: this is easily done by mocking the api functions and ensuring the called variable matching what is expected, any more than this is not needed

File: velo_action/action.py (hunk 1 above)
Comment (RmMsr) [reply velo_action/action.py:71]: What authentication would that be? The velo-action in any case needs to interact with gcp.

File: velo_action/gcp.py (hunk 3 above)
Comment (RmMsr) [reply velo_action/gcp.py:23]: Added a better error message.

For missing credentials: 

> RuntimeError: No valid credentials to access Google cloud. Please either specify the INPUT_SERVICE_ACCOUNT_KEY environment or authenticate using 'gcloud auth login'.


For lack of permissions:

> RuntimeError: Missing permission to access secret 'velo_action_octopus_server' in project 'nube-velo-prod'. Elevate your permissions with: klipy power elevate --group nube.project.editor.nube-velo-prod

File: velo_action/action.py (hunk 1 above)
Comment (andersliland) [reply velo_action/action.py]: https://cloud.google.com/iam/docs/workload-identity-federation

and

https://github.com/marketplace/actions/authenticate-to-google-cloud#authenticating-via-workload-identity-federation

but this requires us to setup the identity-federation first

Comment (RmMsr) [reply velo_action/action.py]: Ok, thanks.

File: velo_action/action.py [hunk 6]
```diff
@@ -44,91 +53,112 @@ def action(input_args: Settings):
...
+
+    if input_args.create_release:
+        logger.info(f"Uploading artifacts to '{velo_artifact_bucket}'")
+        gcloud.upload_from_directory(
+            deploy_folder, velo_artifact_bucket, f"{input_args.project}/{version}"
         )
-        octopus_cli_api_key = g.lookup_data(
-            input_args.octopus_cli_api_key_secret, VELO_PROJECT_NAME
+
+        commit_id = proc_utils.execute_process(
```
Comment (larseen) [inline velo_action/action.py:97]: Use the `GITHUB_SHA` env variable and just make it shorter
https://docs.github.com/en/actions/learn-github-actions/environment-variables

**Comment by andersliland** [comment]
resolves https://github.com/kolonialno/cloud-platform-tasks/issues/301

**Review by andersliland — APPROVED** [review]
Made some lint and refactor changes.

Removed the process_utils as this was not used. Made an explicit function for generating version.

File: velo_action/settings.py [hunk 7]
```diff
@@ -1,7 +1,6 @@
-# type: ignore
```
Comment (RmMsr) [inline velo_action/settings.py:1]: Ahh, that's why the properties did not show up in the linter. I overlooked that.

**Pull request merged**